---
title: "accreta_RNASeq"
author: "Kristen Wells"
date: "7/25/2019"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Background
This documents the analysis of the accreta transcriptomics RNA seq data set.
Below is code and output from the analysis

Set up with packages and data
```{r}
library(corrplot)
library(DESeq2)
library(RColorBrewer)
library(pheatmap)
library(mTEC.10x.data)
base_path <- "/Users/kristen/Documents/sshfs_dir/"

# Path to the count file
count_file <- paste0(base_path, "analysis_outs/accreta_countTable.txt")

# Path to the sample info
sample_file <- paste0(base_path, "sample_info.csv")

# Load in the files
count_table <- read.table(count_file, sep = "\t", row.names = 1, header = TRUE)
sample_table <- read.table(sample_file, sep = ",", header = TRUE,
                           row.names = 1)


# Make names more legible
count_table_names <- sample_table$Group.Name.B
names(count_table_names) <- rownames(sample_table)
new_colnames <- count_table_names[colnames(count_table)]

count_table_groupb <- count_table
colnames(count_table_groupb) <- new_colnames
```


## Correlation between samples

First, I looked at correlation between samples to determine if there is any
structure to the data. Note, this correlation is raw reads, so it is just
an overview of the data.

```{r}
correlation_matrix_groupb <- cor(count_table_groupb)

corrplot(as.matrix(correlation_matrix_groupb),
         method = "circle", order = "hclust")

correlation_matrix <- cor(count_table)

corrplot(as.matrix(correlation_matrix),
         method = "circle", order = "hclust")
```

From these plots, a few major things jump out.

First, the "bad" samples do seem to stand out from the rest of the samples. This means that their gene expression patterns are different than the rest of the samples. Likely these samples will not be used in further analysis.

Second, we see large groups of samples that have high correlation. This to me is consistent with the findings of few differentially expressed genes between samples, but again, this is just a hunch from the first view of the data.

## Make a DESeq2 object

```{r}
count_table <- count_table[, rownames(sample_table)]

# This makes a DESeq2 object where the count data is our count matrix
# and the colData is our sample data. I say we design based on 
# Group Name B, but this can be altered.
dds <- DESeqDataSetFromMatrix(countData = count_table,
                              colData = sample_table,
                              design = ~Group.Name.B)

# Here, we get rid of all genes that aren't expressed.
keep <- rowSums(counts(dds)) >= 10
dds <- dds[keep,]

dds <- DESeq(dds)

# This transforms the counts to have constant variance.
vsd <- vst(dds, blind=FALSE)
```

## Check data quality

With the normalized, transformed counts, we can look more closely to see if any of the samples are not worth keeping.

```{r}
# This determines distance between samples. It is very simlar to the 
# correlation done above.
sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$Group.Name.B, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

```

Again, The "bad" samples come out as being very different from the other samples, I would recommend removing these from the rest of the analysis.

Also, not that most of the samples seem very similar. I don't see a strong preference for clustering of samples from a specific condition.

```{r}
# Make a PCA plot
pcaData <- plotPCA(vsd, intgroup="Group.Name.B", returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot2::ggplot(pcaData, ggplot2::aes(PC1, PC2, color = Group.Name.B)) +
  ggplot2::geom_point(size = 3) +
  ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggplot2::coord_fixed()

```

The colors aren't great, but again you can see that the "bad" labels are separating out.

To make this more clear:

```{r}
pcaData$quality <- "good"
pcaData$quality[grepl("bad", pcaData$Group.Name.B)] <- "bad"

ggplot2::ggplot(pcaData, ggplot2::aes(PC1, PC2, color = quality)) +
  ggplot2::geom_point(size = 3) +
  ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggplot2::coord_fixed()
```

We will remove these samples for the rest of the analysis

```{r}
to_remove <- grep("bad", colData(dds)$Group.Name.B)
dds_good <- dds[, -to_remove]

# We now need to repeat the DESeq function
dds_good$Group.Name.B <- droplevels(dds_good$Group.Name.B)

dds_good$animal <- droplevels(dds_good$animal)
design(dds_good) <- formula(~ Group.Name.B)
dds_good <- DESeq(dds_good)
vsd_good <- vst(dds_good, blind=FALSE)
```

Now we can start looking for DE genes

```{r}
# Function to return sig genes for a comparison
get_de <- function(object, column, var1, var2){
  res <- DESeq2::results(object, contrast = c(column, var1, var2))
  resOrdered <- res[order(res$pvalue),]
  sigGenes <- res[!is.na(res$padj),]
  sigGenes <- sigGenes[sigGenes$padj < 0.1,]
  return(sigGenes)
}

compare_to <- "WT"
compare_against <- c("Carp_L", "diL_good", "estLgood", "noCarp_L", "WT_R")
all_sig <- sapply(compare_against, function(x) get_de(dds_good, "Group.Name.B",
                                                      compare_to, x))
```
```{r}

print(all_sig)
```

```{r}
compare_to <- "diL_good"
compare_against <- c("estLgood")
all_sig <- sapply(compare_against, function(x) get_de(dds_good, "Group.Name.B",
                                                      compare_to, x))
print(all_sig)
```

```{r}
# Make a PCA plot
pcaData <- plotPCA(vsd_good, intgroup="Group.Name.B", returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))

ggplot2::ggplot(pcaData, ggplot2::aes(PC1, PC2, color = Group.Name.B)) +
  ggplot2::geom_point(size = 3) +
  ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggplot2::coord_fixed() +
  ggplot2::scale_color_brewer(palette = "Set1")
```

Changing what colors the PCA can help detect patterns

```{r}
# Change some colors in the PCA plot
left <- c("Carp_L", "diL_good", "estLgood", "noCarp_L", "WT")
right <- c("diR_good", "estRgood", "noCarp_R", "WT_R")
pcaData$rvl <- "no_val"
pcaData$rvl[pcaData$Group.Name.B %in% left] <- "left"
pcaData$rvl[pcaData$Group.Name.B %in% right] <- "right"
carp <- c("Carp_L")
di <- c("diL_good", "diR_good")
est <- c("estRgood", "estLgood")
noCarp <- c("noCarp_L", "noCarp_R")
wt <- c("WT", "WT_R")
pcaData$trt <- "no_val"
pcaData$trt[pcaData$Group.Name.B %in% carp] <- "carp"
pcaData$trt[pcaData$Group.Name.B %in% di] <- "di"
pcaData$trt[pcaData$Group.Name.B %in% est] <- "est"
pcaData$trt[pcaData$Group.Name.B %in% noCarp] <- "noCarp"
pcaData$trt[pcaData$Group.Name.B %in% wt] <- "wt"

ggplot2::ggplot(pcaData, ggplot2::aes(PC1, PC2, color = rvl)) +
  ggplot2::geom_point(size = 3) +
  ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggplot2::coord_fixed() +
  ggplot2::scale_color_brewer(palette = "Set1")

ggplot2::ggplot(pcaData, ggplot2::aes(PC1, PC2, color = trt)) +
  ggplot2::geom_point(size = 3) +
  ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  ggplot2::coord_fixed() +
  ggplot2::scale_color_brewer(palette = "Set1")
```

We colored by treatment and by right vs left. I don't see any obvious patterns here. I suspect that animal may be what separates them.

```{r}
sample_table_new <- sample_table[-to_remove, ]
if(all.equal(rownames(pcaData), rownames(sample_table_new))){
  pcaData$animal <- sample_table_new$animal
  ggplot2::ggplot(pcaData, ggplot2::aes(PC1, PC2, color = animal)) +
    ggplot2::geom_point(size = 3) +
    ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
    ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
    ggplot2::coord_fixed()
}
```

This does seem to be true. So I would say that here the greatest variation is between animals and not between conditions. This is likely why you aren't getting DE genes.

```{r}
sample_table_new$sample_name <- 
  sub("_.*", "", sample_table_new$Sample.Name.in.Report)
if(all.equal(rownames(pcaData), rownames(sample_table_new))){
  pcaData$sample_name <- sample_table_new$sample_name
  ggplot2::ggplot(pcaData, ggplot2::aes(PC1, PC2, label = sample_name)) +
    ggplot2::geom_point(size = 1) +
    ggplot2::xlab(paste0("PC1: ", percentVar[1], "% variance")) +
    ggplot2::ylab(paste0("PC2: ", percentVar[2], "% variance")) +
    ggplot2::coord_fixed() +
    ggrepel::geom_text_repel(ggplot2::aes(label=sample_name),hjust=0, vjust=1,
                       size = 3)
}
```


Now we can look for genes that correlate with the PCs

```{r}
library(factoextra)
rv <- rowVars(assay(vsd_good))
select <- order(rv, decreasing = TRUE)[seq_len(min(200, 
                                                   length(rv)))]
pca <- prcomp(t(assay(vsd_good)[select, ]))
fviz_pca_ind(pca,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
)

fviz_pca_var(pca,
              col.var = "contrib",
              repel = TRUE,
              select.var = list(contrib = 15))

```

Now we can see how much these genes contribute to PCs
```{r}
fviz_contrib(pca, choice = "var", axes = 1, top = 20)
fviz_contrib(pca, choice = "var", axes = 2, top = 20)
```

Get list of genes

```{r}
gene_vals <- data.frame(pca$rotation)
genes <- gene_to_ensembl$V2
names(genes) <- gene_to_ensembl$V1

get_genes <- function(pca_rotation, PC, up = TRUE) {
  if(up){
    pc_order <- gene_vals[order(-gene_vals[[PC]]),]
  } else {
    pc_order <- gene_vals[order(gene_vals[[PC]]),]
  }
  gene_list <- rownames(pc_order[1:100,])
  gene_list_gene <- genes[gene_list]
  gene_list_gene <- unname(gene_list_gene)
  return(gene_list_gene)
}

pc1_up <- get_genes(gene_vals, "PC1")
pc1_down <- get_genes(gene_vals, "PC1", up = FALSE)
pc2_up <- get_genes(gene_vals, "PC2")
pc2_down <- get_genes(gene_vals, "PC2", up = FALSE)
```

Genes pulling to the left on PC1 (towards the WT samples)

```{r}
print(pc1_up)
```

Genes pulling to the right on PC1 (away from the WT samples)

```{r}
print(pc1_down)
```

Genes pulling up on PC2

```{r}
print(pc2_up)
```

Genes pulling down on PC2

```{r}
print(pc2_down)
```